### PR TITLE
Add support for include_str

### PR DESCRIPTION
**Pull request by lnicola**
r? @edwin0cheng

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Maybe we could reuse https://github.com/rust-analyzer/rust-analyzer/blob/6a067ce947980d887a254d71d7183c3d306d8978/crates/rust-analyzer/tests/heavy_tests/main.rs#L468 and add a test ?

**Comment by lnicola** [comment]
I'm not sure what to test for? For some reason, "expand macros recursively" doesn't seem to work.

**Comment by edwin0cheng** [comment]
Adding something like this inside the `out_dir_checks` should works: 
```
let should_<|>be_str = include_str!(“text.rs”);
```

**Comment by lnicola** [comment]
For some reason, we don't seem to find the file -- `relative_file(db, arg_id.into(), &path)` is `None` even though `path` looks correct. This happens even with `include!("main.rs")`, but we might not find the file in the database anyway. Should we expand to `""` for now?

**Comment by edwin0cheng** [comment]
iirc, we added a constraint for limiting self including in `relative_path` in `include!` case. Maybe we should add an enum to turn off it if needed ?

**Comment by lnicola** [comment]
Oh, right, it works with other files. But what about non-`.rs` files? I assume they're not going to be found in the database.

**Comment by edwin0cheng** [comment]
I remember we have some discussion with @matklad that all non-rs files should be included by users. So I think it is okay for now

**Comment by lnicola** [comment]
Added a small hack.

File: crates/rust-analyzer/tests/heavy_tests/main.rs [hunk 1]
```diff
@@ -523,32 +525,40 @@ fn main() {
...
     });
     assert!(res.to_string().contains("&str"));
+    let res = server.send_request::<HoverRequest>(HoverParams {
+        text_document_position_params: TextDocumentPositionParams::new(
+            server.doc_id("src/main.rs"),
+            Position::new(20, 10),
+        ),
+        work_done_progress_params: Default::default(),
+    });
+    assert!(dbg!(res.to_string()).contains("&str"));
```
Comment (edwin0cheng) [inline crates/rust-analyzer/tests/heavy_tests/main.rs:540]: Remain `dbg` ?

**Comment by edwin0cheng** [comment]
Just small nits, otherwise LGTM

bors d+

**Comment by lnicola** [comment]
> Just small nits, otherwise LGTM

Which nits? :smile:

**Comment by lnicola** [comment]
bors r=edwin0cheng

**Pull request merged**